### DISCUSSION TITLE
[Proposal] A system for tracking ALL reads and writes to mutable state (including I/O)

### DISCUSSION THREAD

**Discussion by nmsmith** [modular/modular#368] [Ideas]
In this post, I propose a system for tracking all reads and writes to mutable state within a Mojo program, including I/O. This proposal is a response to recent discussions about "pure functions", such as #345. It aims to achieve the _benefits_ of pure functions (i.e. local reasoning), without actually needing to prohibit mutation.

I hope that my proposed system will provide the following benefits:
- It will allow Mojo programmers to reason about the effects of function calls, thus making it easier to write correct programs.
- It will allow the Mojo _compiler_ to reason about the effects of function calls, thus making Mojo programs easier to optimize and parallelize.
- It will allow Mojo's borrow-checker to be more intelligent, thereby making it easier for Mojo programmers to write memory-safe programs.

These claims are substantiated in the sections that follow.

This proposal is merely a first draft. It is the result of ~2 days of design work. Go easy on me 🙂.

## The relationship between I/O and mutable state

Consider the following program, written in a typical programming language (i.e. not specifically Mojo):

```
var text = ""

fn print(s: String):
    text.append(s)

fn double(x: Int) -> Int:
    print("hello")
    return x * 2

fn triple(x: Int) -> Int:
    print("world")
    return x * 3

let value = double(4) * triple(3) * triple(3)
```

In most programming languages, this is a perfectly valid program. `double` and `triple` are functions that return a value, and also perform a side-effect. Namely, they `print` something. In doing so, they mutate a piece of remote state. (By "remote", I just mean a piece of state that is far away from the function's definition.)

I chose the name `print` to illustrate a point. In the above program, `print` is not doing I/O. But to the best of my knowledge, all I/O can be understood as reading and/or mutating a piece of remote state. For example, one might be reading or mutating a file, or receiving (=reading) and sending (=mutating) messages over a network. In these cases, just imagine the variable `text` in the above program is replaced with a handle to a remote resource.

The problem with the above program is that the programmer (and compiler) cannot reason about the remote mutation. When the programmer reads the signature of the functions `double` and `triple` (which in general, might be imported from a library), all they can tell is that the functions accept an integer, and return an integer. Thus, from the programmer's perspective, these functions seem to be "pure" mathematical functions. The only way to tell what the functions _really_ do is to dig into their implementation, which might not even be available.

The inability to reason about the side-effects of a function call has significant real-world consequences. For example, consider the last line of the above program:
```
let value = double(4) * triple(3) * triple(3)
```

A programmer that sees this code might decide that the two calls to `triple` are redundant, and should be refactored into one call:
```
let t = triple(3)
let value = double(4) * t * t
```

However, this refactoring has significantly changed the behaviour of the program. In the original program, the side-effects of the three function calls result in the variable `text` having the value `"helloworldworld"`. But in the refactored program, `text` will end up having the value `"worldhello"`. Oops!

This is not just an issue for the programmer. The _compiler_ might wish to perform the same refactoring! But the compiler is unable to do this safely, unless it performs an analysis of the entire chain of dependencies of each function definition. (But even then, if actual I/O is being performed, the compiler cannot make any assumptions about the effect of reordering the I/O calls, unless the programmer provides information about their interdependency.)

For Mojo, there is an additional incentive for having a deep understanding of side-effects: Mojo aims to be a programming language for massively-parallel computing! But a programmer cannot safely parallelize a program (and a compiler cannot safely **auto-parallelize** a program) unless the programmer/compiler is aware of the side-effects of each function call!

In the remainder of this post, I want to present a vision for how Mojo can track reads and writes to remote state, such that both the programmer and the compiler can understand when it is safe to reorder, deduplicate, or parallelize the invocation of functions!

## Tracking reads and writes to non-local variables

In the previous section, I argued that remote mutation prevents the programmer and compiler from reasoning about program behaviour. In this section, I propose a solution to that problem.

In the program I presented earlier, the ultimate source of side-effects was the `print` function. The `double` and `triple` functions "inherited" the side-effect of `print` as a consequence of invoking it. Thus, side-effects are "contagious", just like `async` and checked exceptions.

Correspondingly, I propose that a function's side-effects be placed in its type signature. Here is what this would look like for the program presented earlier:

```
var text = ""

fn print(s: String){inout text}:
    text.append(s)

fn double(x: Int){inout text} -> Int:
    print("hello")
    return x * 2

fn triple(x: Int){inout text} -> Int:
    print("world")
    return x * 3

let value = double(4) * triple(3) * triple(3)
```

The only change is the addition of curly braces after the argument list of each function definition. Inside these curly braces is the **non-local variable** that each function mutates. Namely, the variable `text`. Important points:
- Inside the curly braces, the word "text" does not define a new variable. Instead, it refers to an _existing_ variable defined in an outer scope. Such variables are known as "non-local variables".
- The keyword `inout` indicates that the non-local variable will be _mutated_ when each function is invoked. If the non-local variable will only be _viewed_, then a different keyword should be used. (At the time of writing, Mojo's keywords haven't been finalized yet. Thus, at the time you are reading this, the keyword `inout` might not seem appropriate. Apologies! Mentally replace `inout` with your preferred keyword.)
- In the above program, each function only references **one** non-local variable. But in general, a function can reference arbitrarily-many non-local variables. In that case, you would write them like you would write a Python set: `{inout a, inout b...}`.

I will henceforth refer to the variables within the curly braces as the "non-locals" of a function, in analogy to "parameters" and "arguments". You can read the signature of `double` as follows:
> The function `double` accepts an argument `x` and the non-local `text`.

The terminology that I am using in this proposal is just for the purpose of exposition. I am not pushing for the adoption of any particular terminology.

### How these annotations enable local reasoning

Look at the new type signatures for `double` and `triple`:

```
fn double(x: Int){inout text} -> Int:
fn triple(x: Int){inout text} -> Int:
```

From these type signatures, we can learn two things:
- Neither function is "idempotent", in the sense that executing the function multiple times has the same effect as executing it once. (Because successive calls cause successive mutations of `text`.)
- The order in which `double` and `triple` are executed matters, because they both mutate the same non-local variable.

In other words... recall that refactoring we discussed earlier:
```
# Before:
let value = double(4) * triple(3) * triple(3)

# After:
let t = triple(3)
let value = double(4) * t * t
```

The new type signatures of `double` and `triple` make it abundantly clear that this refactoring will change the semantics of the program. Nice! Accordingly, the programmer and compiler should not attempt to "optimize" the program by making this change. (That is, unless the programmer understands and approves of the changed semantics.)

## About the burden of annotation

At this point, you may have the following question:
> Isn't it going to be impractical to list all of the non-local variables that each function reads and/or mutates? For example, if I am doing `print`-debugging, will I have to specify that each of my functions mutates the standard output stream? e.g. `fn foo(){inout stdout}`.

Yes, I agree that this would be too much of a burden in practice — especially when one needs to quickly drop a `print` somewhere. If you do such a thing, you shouldn't be forced to annotate the entire chain of function calls leading up to the `print` statement!

I think a really good solution to this problem is just to have the compiler _infer_ each function's non-locals. This information can then be shown to users via IDE tooltips and [decorations](https://vscode.rocks/decorations/). That said, programmers should probably be encouraged to explicitly annotate the non-locals of functions that are exported from modules. The reason: these non-locals are part of the module's API! If the functions exported by a library are modified such that they reference additional non-locals, this is a breaking change to the library.

## Big bonus: A more expressive borrowing system!

So far, I have been focusing on how explicitly tracking non-local variables can make it easier to reason about a program's behaviour. However, there is another major benefit: it allows us to make Mojo's borrow-checking rules more generous, while upholding memory safety!

Consider a variation upon the earlier program:

```
var text = ""

fn print(s: String){inout text}:
    text.append(s)

fn println(s: String){inout text}:
    text.append(s)
    text.append("\n")

print("hello")
println("world")
```

Such a program cannot type-check in Rust. First of all, named functions in Rust (for some strange reason) cannot refer to non-local variables. So, we might attempt to write the program using anonymous functions (which in Rust, act as closures):
```
let mut text = String::new();

let mut print = |s: &str| {
    text.push_str(s);
};
let mut println = |s: &str| {
    text.push_str(s);
    text.push_str("\n");
};
print("hello");
println("world");
```

But when you attempt to compile this program, you get a borrow-checking error, which I shall paraphrase:
> Upon its creation, the `print` closure mutably borrows `text`. Thus, the `println` closure cannot _also_ borrow `text`.

The issue is that Rust closures borrow non-local variables at the moment they are created (rather than the moment they are invoked). Thus it is not possible for multiple closures to describe the mutation of the same non-local.

However, Rust offers an escape hatch. One can use a `RefCell` to trick the Rust compiler into believing that the closures only need to borrow `text` _immutably_. This allows both of the closures to borrow it. Then, the closures can invoke the `borrow_mut()` method on the `RefCell`, which exposes a mutable reference. The `RefCell` checks _dynamically_ that nobody else has borrowed the value:

```
let text = RefCell::new(String::new());

let print = |s: &str| {
    text.borrow_mut().push_str(s);
};
let println = |s: &str| {
    text.borrow_mut().push_str(s);
    text.borrow_mut().push_str("\n");
};
print("hello");
println("world");
```

The issue with `RefCell` (other than the syntactic noise) is that it provides no compile-time assurances. If you use it incorrectly, your program will crash with a runtime error. I believe that Mojo can do better than this!

Let's look again at the (hypothetical) Mojo version of this program:

```
var text = ""

fn print(s: String){inout text}:
    text.append(s)

fn println(s: String){inout text}:
    text.append(s)
    text.append("\n")

print("hello")
println("world")
```

Here, we have annotated `print` and `println` with the non-local variable (`text`) that they need to borrow. In contrast to the Rust program, we are not declaring that the functions must borrow `text` _immediately_. Instead, we are declaring that the functions must borrow `text` at the moment they are _invoked_. In the above program, both invocations occur at a time when `text` is available to borrow. Thus, the compiler can see that program is safe to execute. No `RefCell` is necessary!

In contrast, imagine the following line is appended to the program:
```
print(text)
```

Here, we are attempting to have `print` borrow `text` immutably, as an argument. However, `print` also needs to borrow `text` _mutably_ (per its type signature). Thus, the program would fail with a borrow-checking error. The fact that this program fails to compile is a **really good thing**, because if it _were_ to compile, it would ultimately execute `text.append(text)`, which (depending on how `append` is implemented) could result in a use-after-free error!

Let's consider a more complicated situation: How should the borrowing of non-locals behave when functions are passed as arguments to _other_ functions?

Here is a simple rule that (to the best of my knowledge) guarantees memory safety:
> Whenever a function `f` is used as an argument to a function `g`, all of `f`'s non-locals are borrowed _immediately_, before `g` is invoked. 

This allows the compiler to reject a bogus situation where `g` believes that it has exclusively borrowed a non-local, but then it invokes a function `f` passed as an argument, which secretly mutates the same non-local:
```
var s = ""

fn f(){inout s}:
    s.append("hello")

fn g(func: fn()){inout s}:
    inout b = s  # Borrow s (a hypothetical syntax for local refs)
    func()
    print(b)     # Use s

g(f)  # Error: f borrows s when passed as an argument. Thus, g is unable to borrow s as required.
```
In short: the compiler rejects any program that invokes `g` in a context where `s` is not available to borrow.

But we can be more flexible than this. We can allow `g` to accept a function that might borrow the non-local `s`:
```
fn g(func: fn(){inout s}){inout s}:
```
If `g` has the above type signature, then it _can_ borrow the same non-local that `func` borrows, as long as the borrows don't overlap!
```
fn g(func: fn(){inout s}){inout s}:
    func()    # "func" borrows s and returns it.
    print(s)  # Now s can be borrowed again.

g(f)  # Now this compiles!
```

In conclusion: I think this proposed semantics for borrowing non-locals would be a lovely extension to Mojo's existing borrowing system, allowing more programs to be compiled _safely_, and without requiring escape hatches analogous to Rust's `RefCell`.

## Modelling the private (i.e. hidden) state of libraries

A concern that I have not yet addressed is how to handle functions that are imported from another module (which might belong to an external library). If a function that is imported from another module references non-locals that belong to the _private state_ of the module, what should the type signature of the function look like?
- **Note:** I am assuming that Mojo modules can be stateful—that they behave like singleton objects. (Maybe even: a module can be instantiated multiple times.) If modules _cannot_ be stateful, then the following discussion is moot.

Consider the following import statement:
`from Library import foo, bar`

Suppose that `Library` contains some private state—perhaps a `log` variable and a `cache` of some kind. Furthermore, suppose that these variables are non-locals of `foo` and `bar`. When `foo` and `bar` are imported, the user of the library should not be exposed to these internal details. Despite this, the user needs to know that invoking these functions mutates _some_ non-local. Otherwise, they would experience the reasoning challenges discussed at the beginning of this document: they would not know whether calling `foo` twice has the same effect as calling it once, or whether calls to `foo` and `bar` can be reordered.

To address this issue, the library can export `foo` and `bar` as follows:
```
export foo(...){inout this}:    # This is a hypothetical syntax for exports.
export bar(...){inout this}:
```
Here, "this" is a keyword that is being used to refer to the enclosing module. (Perhaps another keyword would be better, but let's not worry about that right now.)

If `foo` and `bar` are exported as per above, then when the user imports the functions, they will see the following type signatures:
```
foo: fn(...){inout Library}
bar: fn(...){inout Library}
```

Here, the _whole module_ is being used as a non-local. This informs the programmer that the order of every call to `foo` and `bar` is potentially significant.

This approach gives library authors two much-needed affordances:
- It allows library authors to hide information about the non-locals that are being mutated.
- It allows library authors to make their library _forward-compatible_ with future API changes, by making _no promises_ that calls to the library can be reordered.

This approach is highly analogous to how `inout Self` is used in the signatures of `struct` methods. Basically, the module is being treated as an opaque object, and the functions that it exports are being treated as methods that (potentially) mutate the object.

## Reasoning about the mutation of class instances (interfacing with Python)

I have claimed that annotating functions with their non-locals allows the programmer and compiler to know whether, for any pair of function calls, the order in which they are invoked is significant. This relies on a crucial property of Mojo structs: a variable that refers to a mutable struct has **no other aliases.**

Unfortunately, objects (i.e. `class` instances) do not have this property. Objects are heap-allocated and garbage-collected, making it "safe" (under a generous interpretation of "safe") for a program to contain many mutable references to the same object. As a consequence, it is typical for an object-oriented program to consist of a complex web of aliased mutable state. Consequently, it does not seem possible—in the general case—to know which objects might be mutated by a particular function call. This difficulty is compounded by the fact that Python code is usually untyped (or "unityped"), so one may not even know which _classes_ of objects will be mutated by a particular function call.

Given these facts, how can the mutation of objects be integrated into my proposal? Well, as a bare minimum, we could introduce a _pseudo_ non-local that refers to the set of all objects. For the time being, let's call this non-local "`objects`". Any function that invokes a method upon an object (or rebinds a non-local variable that refers to an object) would have `{inout objects}` added to its type signature.
- For Python's built-in data types, we can probably be more precise. For example, Python integers/floats/bools/strings are immutable. Thus, method calls that operate solely on such objects cannot change the program's state.

The approach that I am proposing here is very preliminary. To reach a final design, we would need to come up with a comprehensive characterization of the state that is able to read/modified after an arbitrary `def` call, and after calls to each function/method that is built into Python. (Unfortunately, I don't have time to work on this right now, but I hope to have the opportunity eventually.)

Also, it's worth nothing that Mojo's borrow checker must ignore `{inout objects}` annotations, because it must be possible to borrow/mutate `objects` multiple times simultaneously. The sole purpose of these annotations is therefore to inform the programmer and compiler that a function call reads or mutates one or more Python objects (i.e. that it has a side-effect), and therefore reordering such calls could potentially change the program's behaviour.
- Given this inconsistency with standard the borrowing rules for `inout`, perhaps it would be better to use a different keyword. Anything that means "I will read and/or modify arbitrary objects" would be appropriate.

So, what effect would this proposal have upon a Mojo program that makes heavy use of Python libraries? Well, functions that manipulate objects (i.e. all the functions imported from Python libraries, and all the functions that depend on those functions) would inherit the `{inout objects}` non-local. Remember though: I have proposed that all non-locals be automatically inferred (except for functions that are part of a library's API), thus, I wouldn't expect the contagious spread of `{inout objects}` to be an issue. Effectively, the annotation is just auto-generated _documentation_, letting users know that the code interacts with Python objects, and that the effect of the call upon the program's state cannot be precisely characterized.

One other thing that needs to be considered is how/whether to track the I/O operations that are performed by a Python library. I haven't put much thought into that issue yet.

## Summary

I've presented a rough sketch of a system for comprehensively tracking all reads and writes to the state of a Mojo program. I discussed several benefits that this system will hopefully provide:
- It will allow Mojo programmers to reason about the effects of function calls, thus making it easier to write correct programs.
- It will allow the Mojo _compiler_ to reason about the effects of function calls, thus making Mojo programs easier to optimize and parallelize.
- It will allow Mojo's borrow-checker to be more intelligent, thereby making it easier for Mojo programmers to write memory-safe programs.

My goal was to introduce the seed of a potentially promising idea. Let me know what you all think! 🙂

**Comment by mzaks** [comment]
Thank you for the very detailed proposal, it was really nice to read and is a very good way of understand your trail of thought.

I think there is a general issue with allowing of borrow non-local variables as there are scoping and life time issues. 

As you mentioned, Rust solves it by forcing us to use RefCell, which might lead to runtime errors.

Here is a nice example how such a panic can be produced:
[RefCell panic example](https://users.rust-lang.org/t/how-refcell-could-result-in-thread-panic/38715/3)

Do you think with you proposal, it would be possible to catch this behaviour at compile time?

**Reply by nmsmith** [reply]
Yes, I hope it is possible to come up with a static semantics for borrowing non-locals that makes runtime borrowing errors and memory errors impossible. More investigation is required, but the examples I’ve given suggest that there is a good chance that Mojo can achieve this.

**Comment by lattner** [comment]
I've been overwhelmed by other things this week, but hope to review this this weekend I'm sorry for the delay - thank you for putting it together!

**Comment by strangemonad** [comment]
@nmsmith this is great, I'm still digesting bits of it. I'd suggest that an additional benefit seems like it could help reason about certain side effects under const-evaluation e.g. printing or reading files during parameter const-eval eval time vs run time. Being able to read external IO during const eval can lead to powerful expressivity at compile time:
- reading a remote DB's sql schema or http API schema let's you build type-safe DB and API access (see F# and C# type providers)
- You could imagine reading a GPU execution cost model at const-eval time to be used in auto-tune e.g. you could leverage GPU cost models similar to the work done by Vivienne Sze https://arxiv.org/abs/1703.09039

Aside: I'm being specific about calling it const-eval time vs generally compile time because the mental model that I think could be powerful for defining the semantics of const-eval is inspired by this note by Sterling and Harper https://www.jonmsterling.com/papers/sterling-harper-2021-mlw.pdf. It summarizes a bunch of their work that supports that the single compile time vs runtime phase distinction is just one arbitrary phase distinction out of many and it sketches a framework for how you might track all sorts of parameters to reason about different program aspects (e.g. session types for communicating actors, resource usage, etc).

**Comment by lattner** [comment]
@nmsmith thank you so much for writing up this thoughtful note, it has a strong and well communicated motivation.  Agreed that syntax/keywords are not the important thing. Great that you focus on separate compilation - reasoning about behavior of method based on signature not body.

It seems like you're already thinking about this in terms of compiler inference, but if this is a type system feature, I'd encourage you to think about this as a three level design space:

1) no effect tracking as part of signatures.
2) coarse grained tracking - e.g. "side effect free" or "only reads side effects" like the pure/const attribute gcc extensions
3) fine grained tracking, like your proposal starts with.

In my opinion, you can't do #3 without also doing #2.  The reason for this is that you have to support "undescribable side effects" - this is necessary because you have things like inline assembly (which can change machine status registers), mojo can have arbitrary unknown MLIR operations, imported Python can do anything, etc.  Given this, we can focus on #2.  There is a huge design space for #3, and while we can debate the tradeoffs, the issues with #2 are the core issue.

Your proposal identifies the challenges with #2, including the need to add printfs to places.  I'd also say that imported Python and dynamically typed code is core to Mojo and as such, virtually everything will have to be treated as having side effects in those cases.

Going further, there are bigger problems with this being in the type system by making it part of a function signature: For example,
 -  this will interfere significantly with traits, because traits will want to abstract over multiple implementations of a function, and those functions will have different effect settings.  I don't believe that (in practice) you'll have traits that can have useful effects - they'll have to declare they can do anything - because we want code implemented with untyped bodies to participate.
 -  the proposed inference rule will interfere with separate compilation and upgradability of packages.  Adding a print to a function would transitively taint all the functions calling it ABI's, which transitively impacts all the packages built on it.  This would require a full rebuild of everything, and breaks binary distributions of packages.
 - Indirections through classes are a significant issue - there are reasonable ways to implement this (e.g. see how the Swift laws of exclusivity handle untrackable class pointers, same issue) but classes (and things like indirect pointer dereferences) are going to be very common.

Furthermore, It isn't clear that this is complexity we /want/ to foist onto the user :). Your proposed approach of having the compiler infer this and use/propagate this information is a good one - compilers like LLVM already do this - but that doesn't require language support.  The only way we could make take advantage of this for parallelization is if it is part of the type system (see problems with that above), and the "subexpression elimination of triple" example given doesn't require anything fancy - we can already do that at either the LLVM or MLIR level without a language feature.

We can discuss compiler implementation details at some point, and it is super clear to me that we want the MLIR based optimizations in Mojo to be aggressive with memory/side effect optimization, but I don't see how we could possibly bring this into the type system in a practical way.  If you'd like to do that, I'd request that we push it off until after the basic trait system is up and running, because the intersection with that is quite significant.

**Reply by nmsmith** [reply]
Thank you for the detailed feedback Chris! You've certainly helped me understand some of the challenges that would need to be overcome for this proposal to be practical.

I think I've identified a helpful framing of the biggest issue(s). The framing is:
- It is easy to model the side-effects of a concrete function, but it is difficult to **abstract over** the side-effects of a function received as an _argument/parameter_.

A function is usually received as an argument/parameter under one of the following guises:
- A function pointer or closure.
- A trait method. The concrete method is either passed via a hidden argument (dynamic dispatch), or made available through monomorphization (static dispatch).
- An abstract or overridable method defined within a class inheritance hierarchy. (Usually dynamic dispatch.)

When such functions are received as arguments/parameters, it would be very uncommon for users to want to think about _specific_ side-effects that may occur. Instead, we would want a coarse categorization such as the one you mentioned:
- Mutates non-local state. (Maximal side-effects; only commutes with pure functions.)
- Reads non-local state. (Commutes with other function calls of this kind, and with pure functions.)
- Pure function. (Commutes with everything.)

But as you suggested, in most situations, users don't want/need to be thinking about side-effects _at all_—and certainly not each time they define a function that accepts generic arguments.

I believe you're right—it's probably best to shelve this proposal until Mojo has traits and classes.

Thank you for taking the time to respond to my post Chris. Your feedback was very helpful 🙂.

**Reply by lattner** [reply]
Happy to Nick, thank you for investing time to explore this!

**Comment by nmsmith** [comment]
For future reference: Martin Odersky (designer of Scala) has been working on a vaguely similar system ([docs](https://dotty.epfl.ch/docs/reference/experimental/cc.html), [paper 1](https://arxiv.org/pdf/2207.03402.pdf), [paper 2](https://arxiv.org/pdf/2306.06496.pdf)) for tracking side-effects.

**Comment by ivellapillil** [comment]
Nim language has generalized exception tracking through use of effect system (https://nim-lang.org/docs/manual.html#effect-system). It would be probably useful to take a look on how they implement it. It could also be used for tracking "unsafe" code usage.

**Comment by strangemonad** [comment]
The [Poutine effect handler system](https://docs.pyro.ai/en/stable/poutine.html) is also at the heart of how Pyro implements probabilistic programming on top of pytorch
